src/wasmedge/wasmedge.cc

```diff
@@ -0,0 +1,600 @@
+  }
+  return 0;
+}
+
+std::optional<std::string_view> WasmEdge::getMemory(uint64_t pointer, uint64_t size) {
+  char *ptr = reinterpret_cast<char *>(WasmEdge_MemoryInstanceGetPointer(memory_, pointer, size));
+  if (!ptr) {
+    return std::nullopt;
+  }
+  return std::string_view(ptr, size);
+}
+
+bool WasmEdge::setMemory(uint64_t pointer, uint64_t size, const void *data) {
+  auto res = WasmEdge_MemoryInstanceSetData(memory_, reinterpret_cast<const uint8_t *>(data),
+                                            pointer, size);
+  return WasmEdge_ResultOK(res);
+}
+
+bool WasmEdge::getWord(uint64_t pointer, Word *word) {
+  constexpr auto size = sizeof(uint32_t);
+  uint32_t word32;
+  auto res =
+      WasmEdge_MemoryInstanceGetData(memory_, reinterpret_cast<uint8_t *>(&word32), pointer, size);
+  if (WasmEdge_ResultOK(res)) {
+    word->u64_ = word32;
+    return true;
+  }
+  return false;
+}
+
+bool WasmEdge::setWord(uint64_t pointer, Word word) {
+  constexpr auto size = sizeof(uint32_t);
+  uint32_t word32 = word.u32();
+  auto res =
+      WasmEdge_MemoryInstanceSetData(memory_, reinterpret_cast<uint8_t *>(&word32), pointer, size);
+  return WasmEdge_ResultOK(res);
+}
+
+template <typename... Args>
+void WasmEdge::registerHostFunctionImpl(std::string_view module_name,
+                                        std::string_view function_name, void (*function)(Args...)) {
+  auto it = import_objects_.find(std::string(module_name));
+  if (it == import_objects_.end()) {
+    import_objects_.emplace(module_name, std::make_unique<HostModuleData>(module_name));
+    it = import_objects_.find(std::string(module_name));
+  }
+
+  auto data = std::make_unique<HostFuncData>(module_name, function_name);
+  auto *func_type = newWasmEdgeFuncType<std::tuple<Args...>>();
+  data->vm_ = this;
+  data->raw_func_ = reinterpret_cast<void *>(function);
+  data->callback_ = [](void *data, WasmEdge_MemoryInstanceContext *MemCxt,
+                       const WasmEdge_Value *Params, WasmEdge_Value *Returns) -> WasmEdge_Result {
+    auto func_data = reinterpret_cast<HostFuncData *>(data);
+    const bool log = func_data->vm_->cmpLogLevel(LogLevel::trace);
+    if (log) {
+      func_data->vm_->integration()->trace("[vm->host] " + func_data->modname_ + "." +
+                                           func_data->name_ + "(" +
+                                           printValues(Params, sizeof...(Args)) + ")");
+    }
+    auto args = convValTypesToArgsTuple<std::tuple<Args...>>(Params);
+    auto fn = reinterpret_cast<void (*)(Args...)>(func_data->raw_func_);
+    std::apply(fn, args);
+    if (log) {
+      func_data->vm_->integration()->trace("[vm<-host] " + func_data->modname_ + "." +
+                                           func_data->name_ + " return: void");
+    }
+    return WasmEdge_Result_Success;
+  };
+
+  auto *hostfunc_cxt = WasmEdge_FunctionInstanceCreate(func_type, data->callback_, data.get(), 0);
```

Comment:
Should we verify that the call succeeded and that `hostfunc_cxt != nullptr` after the call?

Reply:
Fixed.